shim.c

```diff
@@ -1831,6 +1831,59 @@ debug_hook(void)
 	x = 1;
 }
 
+EFI_STATUS
+check_shim_sbat()
+{
+	unsigned int i;
+	EFI_STATUS efi_status;
+	list_t sbat_var_entries;
+	struct sbat sbat = { 0 };
+	struct sbat_entry *entry = NULL;
+	char *sbat_start = (char *)&_sbat;
+	char *sbat_end = (char *)&_esbat;
+	size_t sbat_size = sbat_end - sbat_start;
+
+	char *sbat_data = AllocatePool(sbat_size + 1);
+	if (!sbat_data) {
+		console_print(L"Failed to allocate SBAT buffer\n");
+		return EFI_OUT_OF_RESOURCES;
+	}
+
+	CopyMem(sbat_data, sbat_start, sbat_size);
+	sbat_data[sbat_size] = '\0';
+
+	efi_status = parse_sbat(sbat_data, sbat_size + 1, &sbat);
+	if (EFI_ERROR(efi_status)) {
+		console_print(L"shim SBAT data not correct: %r\n", efi_status);
+		goto out;
+	}
+
+	console_print(L"shim SBAT data\n");
+	for (i = 0; i < sbat.size; i++) {
+		entry = sbat.entries[i];
+		console_print(L"%a, %a, %a, %a, %a, %a\n",
+		              entry->component_name,
+		              entry->component_generation, entry->vendor_name,
+		              entry->vendor_package_name, entry->vendor_version,
+		              entry->vendor_url);
+	}
+
+	INIT_LIST_HEAD(&sbat_var_entries);
+
+	efi_status = parse_sbat_var(&sbat_var_entries);
+	if (efi_status == EFI_SUCCESS)
+		efi_status = verify_sbat(&sbat, &sbat_var_entries);
+	if (efi_status == EFI_NOT_FOUND)
+		efi_status = EFI_SUCCESS;
+
+out:
+	if (sbat.entries)
+		for (i = 0; i < sbat.size; i++)
+			FreePool(sbat.entries[i]);
+	FreePool(sbat_data);
```

Comment:
It doesn't look like sbat.entries is freed (allocated here https://github.com/rhboot/shim/blob/06ee1d9f8b06150a6ea4c7893325fdfbf8b29182/sbat.c#L99).

There is similar code (not part of this PR) in handle_image() that has the same issue (and that implementation also leaks sbat_data).

Reply:
You are correct about sbat.entries leaking.

I'll also include those memory leak fixes in handle_image() as a part of this series.

Reply:
@chrisccoulson fixed the leaks in this patch and added preparatory fixes in the pull-request for the other issues you pointed out.